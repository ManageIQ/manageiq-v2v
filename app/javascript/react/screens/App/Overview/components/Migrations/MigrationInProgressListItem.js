import React from 'react';
import PropTypes from 'prop-types';
import { Spinner, ListView, Button, Icon, UtilizationBar } from 'patternfly-react';
import EllipsisWithTooltip from 'react-ellipsis-with-tooltip';

import TickingIsoElapsedTime from '../../../../../../components/dates/TickingIsoElapsedTime';
import getPlaybookName from './helpers/getPlaybookName';
import { MIGRATIONS_FILTERS, TRANSFORMATION_PLAN_REQUESTS_URL } from '../../OverviewConstants';
import {
  countFailedVms,
  isWaitingForConversionHost,
  getRequestsOfPlan,
  getIndexedTasksOfPlan,
  aggregateTasks,
  calculateTotalDiskSpace,
  shouldDisableCancelButton
} from './helpers/inProgressHelpers';
import InProgressRow from './InProgressRow';
import ProgressBarTooltip from './ProgressBarTooltip';

const MigrationInProgressListItem = ({
  plan,
  serviceTemplatePlaybooks,
  allRequestsWithTasks,
  reloadCard, // TODO where does this come from, can we rename it?
  redirectTo,
  fetchTransformationPlansAction,
  fetchTransformationPlansUrl,
  isFetchingTransformationPlans,
  isFetchingAllRequestsWithTasks,
  acknowledgeDeniedPlanRequestAction,
  isEditingPlanRequest,
  setMigrationsFilterAction,
  cancelPlanRequestAction,
  isCancellingPlanRequest,
  requestsProcessingCancellation
}) => {
  const { requestsOfAssociatedPlan, mostRecentRequest } = getRequestsOfPlan({ plan, allRequestsWithTasks });
  const waitingForConversionHost = isWaitingForConversionHost(mostRecentRequest);

  const isInitiating = reloadCard || !mostRecentRequest || mostRecentRequest.request_state === 'pending';

  // Plan request state: initiating
  if (isInitiating) {
    return (
      <InProgressRow
        plan={plan}
        additionalInfo={[
          <ListView.InfoItem key="initiating">
            <Spinner size="sm" inline loading />
            {__('Initiating migration. This might take a few minutes.')}
          </ListView.InfoItem>
        ]}
      />
    );
  }

  // Plan request state: waiting for conversion host
  if (waitingForConversionHost) {
    const cancelButtonDisabled = shouldDisableCancelButton({
      requestsProcessingCancellation,
      mostRecentRequest,
      isFetchingTransformationPlans,
      isFetchingAllRequestsWithTasks,
      isCancellingPlanRequest
    });

    const onCancelClick = event => {
      event.stopPropagation();
      cancelPlanRequestAction(TRANSFORMATION_PLAN_REQUESTS_URL, mostRecentRequest.id).then(() =>
        fetchTransformationPlansAction({ url: fetchTransformationPlansUrl, archived: false })
      );
    };

    return (
      <InProgressRow
        plan={plan}
        additionalInfo={[
          <ListView.InfoItem key="waiting-for-host">
            <Spinner size="sm" inline loading />
            <EllipsisWithTooltip style={{ maxWidth: 300 }}>
              {__('Waiting for an available conversion host. You can continue waiting or go to the Migration Settings page to increase the number of migrations per host.') /* prettier-ignore */}
            </EllipsisWithTooltip>
          </ListView.InfoItem>
        ]}
        actions={
          <Button disabled={cancelButtonDisabled} onClick={onCancelClick}>
            {__('Cancel Migration')}
          </Button>
        }
      />
    );
  }

  // Plan request state: denied / no conversion host configured
  if (mostRecentRequest.approval_state === 'denied') {
    const onCancelClick = () =>
      acknowledgeDeniedPlanRequestAction({
        plansUrl: fetchTransformationPlansUrl,
        planRequest: mostRecentRequest
      }).then(() => setMigrationsFilterAction(MIGRATIONS_FILTERS.completed));

    return (
      <InProgressRow
        plan={plan}
        additionalInfo={[
          <ListView.InfoItem key="denied">
            <Icon type="pf" name="error-circle-o" />
            <EllipsisWithTooltip style={{ maxWidth: 300 }}>
              {__('Unable to migrate VMs because no conversion host was configured at the time of the attempted migration.') /* prettier-ignore */}{' '}
              {__('See the product documentation for information on configuring conversion hosts.')}
            </EllipsisWithTooltip>
          </ListView.InfoItem>
        ]}
        actions={
          <Button disabled={isEditingPlanRequest} onClick={onCancelClick}>
            {__('Cancel Migration')}
          </Button>
        }
      />
    );
  }

  // UX business rule: reflect failed immediately if any single task has failed
  // in the most recent request
  const numFailedVms = countFailedVms(mostRecentRequest);

  // UX business rule: aggregrate the tasks across requests reflecting current status of all tasks,
  // (gather the last status for the vm, gather the last storage for use in UX bussiness rule 3)
  const tasksBySourceId = getIndexedTasksOfPlan({ plan, requestsOfAssociatedPlan, mostRecentRequest });
  const {
    numTotalVms,
    numCompletedVms,
    taskRunningPreMigrationPlaybook,
    taskRunningPostMigrationPlaybook
  } = aggregateTasks(tasksBySourceId);

  // UX business rule: reflect the total disk space migrated, aggregated across requests
  const { totalDiskSpace, totalMigratedDiskSpace } = calculateTotalDiskSpace(tasksBySourceId);

  const redirectToPlan = event => {
    event.stopPropagation();
    redirectTo(`/plan/${plan.id}`);
  };
  const baseRowProps = { plan, numFailedVms, numTotalVms, onClick: redirectToPlan };

  // UX business rule: if there are any pre migration playbooks running,
  // or if all disks have been migrated and we have a post migration playbook running, show this instead
  if (
    taskRunningPreMigrationPlaybook ||
    (totalMigratedDiskSpace >= totalDiskSpace && taskRunningPostMigrationPlaybook)
  ) {
    const { pre_service_id, post_service_id } = plan.options.config_info;
    const playbookId = taskRunningPreMigrationPlaybook ? pre_service_id : post_service_id;
    const playbookName = getPlaybookName(serviceTemplatePlaybooks, playbookId);
    return (
      <InProgressRow
        {...baseRowProps}
        additionalInfo={[
          <ListView.InfoItem key="running-playbook">
            <Spinner size="sm" inline loading />
            {sprintf(__('Running playbook service %s. This might take a few minutes.'), playbookName)}
          </ListView.InfoItem>
        ]}
      />
    );
  }

  return (
    <InProgressRow
      {...baseRowProps}
      additionalInfo={[
        <ListView.InfoItem key="migration-progress">
          <div id={`vm-progress-bar-${plan.id}`} className="vm-progress-bar">
            <UtilizationBar
              now={numCompletedVms}
              max={numTotalVms}
              description={__('Migrating data')}
              label={
                <span>
                  <span id="vms-migrated">{sprintf(__('%s of %s VMs'), numCompletedVms, numTotalVms)}</span>{' '}
                  {__('migrated')}
                </span>
              }
              descriptionPlacementTop
              availableTooltipFunction={() => (
                <ProgressBarTooltip id={`available-vm-${plan.id}`} max={numTotalVms} now={numCompletedVms} />
              )}
              usedTooltipFunction={() => (
                <ProgressBarTooltip id={`used-vm-${plan.id}`} max={numTotalVms} now={numCompletedVms} />
              )}
            />
            <div className="active-migration-elapsed-time">
              <Icon type="fa" name="clock-o" />
              <TickingIsoElapsedTime startTime={mostRecentRequest.created_on} />
            </div>
          </div>
        </ListView.InfoItem>
      ]}
    />
  );
};

MigrationInProgressListItem.propTypes = {
  plan: PropTypes.object.isRequired,
  serviceTemplatePlaybooks: PropTypes.array,
  allRequestsWithTasks: PropTypes.array,
  reloadCard: PropTypes.bool,
  redirectTo: PropTypes.func,
  fetchTransformationPlansAction: PropTypes.func,
  fetchTransformationPlansUrl: PropTypes.string,
  isFetchingTransformationPlans: PropTypes.bool,
  isFetchingAllRequestsWithTasks: PropTypes.bool,
  acknowledgeDeniedPlanRequestAction: PropTypes.func,
  isEditingPlanRequest: PropTypes.bool,
  setMigrationsFilterAction: PropTypes.func,
  cancelPlanRequestAction: PropTypes.func,
  isCancellingPlanRequest: PropTypes.bool,
  requestsProcessingCancellation: PropTypes.array
};

export default MigrationInProgressListItem;
