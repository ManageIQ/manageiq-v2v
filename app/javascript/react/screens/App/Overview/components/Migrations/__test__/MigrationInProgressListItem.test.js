import React from 'react';
import { shallow, mount } from 'enzyme';

import MigrationInProgressListItem from '../MigrationInProgressListItem';
import { transformationPlans } from '../../../overview.transformationPlans.fixtures';
import {
  MIGRATIONS_FILTERS,
  TRANSFORMATION_PLAN_REQUESTS_URL,
  FETCH_TRANSFORMATION_PLANS_URL
} from '../../../OverviewConstants';

describe('when the request is denied', () => {
  const plan = { ...transformationPlans.resources[7], transformation_mapping: { transformation_mapping_items: [] } };
  const [deniedRequest] = plan.miq_requests;
  const allRequestsWithTasks = [{ ...deniedRequest, miq_request_tasks: [] }];
  const baseProps = {
    plan,
    allRequestsWithTasks
  };

  test('renders an error view', () => {
    const wrapper = shallow(<MigrationInProgressListItem {...baseProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('clicking the cancel button acknowledges the dismissal and switches the view to completed migrations', async () => {
    const acknowledgeDeniedPlanRequestAction = jest.fn(() => Promise.resolve());
    const setMigrationsFilterAction = jest.fn();
    const wrapper = mount(
      <MigrationInProgressListItem
        {...baseProps}
        acknowledgeDeniedPlanRequestAction={acknowledgeDeniedPlanRequestAction}
        setMigrationsFilterAction={setMigrationsFilterAction}
      />
    );

    wrapper.find('Button').simulate('click');

    await expect(acknowledgeDeniedPlanRequestAction).toHaveBeenCalled();
    expect(setMigrationsFilterAction).toHaveBeenCalledWith(MIGRATIONS_FILTERS.completed);

    wrapper.unmount();
  });
});

describe('if there are no conversion hosts available', () => {
  const plan = { ...transformationPlans.resources[5], transformation_mapping: { transformation_mapping_items: [] } };
  const [request] = plan.miq_requests;
  const allRequestsWithTasks = [{ ...request, miq_request_tasks: [{ conversion_host_id: null }] }];
  const baseProps = {
    plan,
    allRequestsWithTasks,
    requestsProcessingCancellation: []
  };

  test('renders an error view', () => {
    const wrapper = shallow(<MigrationInProgressListItem {...baseProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('clicking the cancel button initiates cancel at the request level and refreshes plans', async () => {
    const cancelPlanRequestAction = jest.fn(() => Promise.resolve());
    const fetchTransformationPlansAction = jest.fn();
    const wrapper = mount(
      <MigrationInProgressListItem
        {...baseProps}
        cancelPlanRequestAction={cancelPlanRequestAction}
        fetchTransformationPlansAction={fetchTransformationPlansAction}
        fetchTransformationPlansUrl={FETCH_TRANSFORMATION_PLANS_URL}
      />
    );

    wrapper.find('Button').simulate('click');

    await expect(cancelPlanRequestAction).toHaveBeenCalledWith(TRANSFORMATION_PLAN_REQUESTS_URL, request.id);
    expect(fetchTransformationPlansAction).toHaveBeenCalledWith({
      url: FETCH_TRANSFORMATION_PLANS_URL,
      archived: false
    });
  });
});
