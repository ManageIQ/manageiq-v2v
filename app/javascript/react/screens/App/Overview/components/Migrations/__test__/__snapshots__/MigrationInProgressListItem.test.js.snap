// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if there are no conversion hosts available renders an error view 1`] = `
<InProgressRow
  actions={
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      onClick={[Function]}
    >
      Cancel Migration
    </Button>
  }
  additionalInfo={
    Array [
      <ListViewInfoItem
        className=""
        stacked={false}
      >
        <Spinner
          className=""
          inline={true}
          inverse={false}
          loading={true}
          size="sm"
        />
        <EllipisWithTooltip
          style={
            Object {
              "maxWidth": 300,
            }
          }
        >
          Waiting for an available conversion host. You can continue waiting or go to the Migration Settings page to increase the number of migrations per host.
        </EllipisWithTooltip>
      </ListViewInfoItem>,
    ]
  }
  numFailedVms={0}
  numTotalVms={0}
  onClick={null}
  plan={
    Object {
      "configVmLength": 2,
      "created_on": "2018-05-01T12:13:50Z",
      "fulfilledOn": "2018-04-06T12:31:30Z",
      "href": "http://localhost:3000/api/service_templates/60",
      "id": "60",
      "miq_requests": Array [
        Object {
          "approval_state": "approved",
          "created_on": "2018-04-06T12:31:30Z",
          "description": "Migration Plan F-0",
          "destination_id": null,
          "destination_type": null,
          "fulfilled_on": "2018-04-06T12:31:30Z",
          "href": "http://localhost:3000/api/service_requests/6000",
          "id": "6000",
          "message": "VM Transformations <SOMETHING>",
          "options": Object {
            "cart_state": "ordered",
            "delivered_on": "2018-04-06T12:49:30Z",
            "dialog": null,
            "initiator": null,
            "requester_group": "EvmGroup-super_administrator",
            "src_id": "60",
            "workflow_settings": Object {
              "resource_action_id": "2507",
            },
          },
          "process": true,
          "request_state": "active",
          "request_type": "transformation_plan",
          "requester_id": "1",
          "requester_name": "Administrator",
          "service_order_id": "91",
          "source_id": "60",
          "source_type": "ServiceTemplate",
          "status": "Ok",
          "tenant_id": "1",
          "type": "ServiceTemplateTransformationPlanRequest",
          "updated_on": "2018-04-06T12:31:30Z",
          "userid": "admin",
        },
      ],
      "name": "Migration Plan F-0",
      "options": Object {
        "config_info": Object {
          "actions": Array [
            Object {
              "vm_id": "1",
            },
            Object {
              "vm_id": "3",
            },
          ],
          "transformation_mapping_id": "1",
        },
      },
      "scheduleTime": null,
      "status": "Ok",
      "transformation_mapping": Object {
        "transformation_mapping_items": Array [],
      },
    }
  }
/>
`;

exports[`when the request is denied renders an error view 1`] = `
<InProgressRow
  actions={
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      onClick={[Function]}
    >
      Cancel Migration
    </Button>
  }
  additionalInfo={
    Array [
      <ListViewInfoItem
        className=""
        stacked={false}
      >
        <Icon
          name="error-circle-o"
          type="pf"
        />
        <EllipisWithTooltip
          style={
            Object {
              "maxWidth": 300,
            }
          }
        >
          Unable to migrate VMs because no conversion host was configured at the time of the attempted migration.
           
          See the product documentation for information on configuring conversion hosts.
        </EllipisWithTooltip>
      </ListViewInfoItem>,
    ]
  }
  numFailedVms={0}
  numTotalVms={0}
  onClick={null}
  plan={
    Object {
      "configVmLength": 2,
      "created_on": "2018-05-01T12:13:50Z",
      "fulfilledOn": "2018-04-06T12:31:30Z",
      "href": "http://localhost:3000/api/service_templates/80",
      "id": "80",
      "miq_requests": Array [
        Object {
          "approval_state": "denied",
          "cancelation_status": null,
          "created_on": "2018-04-06T12:31:30Z",
          "description": "Migration Plan H-0",
          "destination_id": null,
          "destination_type": null,
          "fulfilled_on": "2018-04-06T12:31:30Z",
          "href": "http://localhost:3000/api/service_requests/8000",
          "id": "8000",
          "message": "VM Transformations <SOMETHING>",
          "options": Object {
            "cart_state": "ordered",
            "delivered_on": "2018-04-06T12:49:30Z",
            "dialog": null,
            "initiator": null,
            "requester_group": "EvmGroup-super_administrator",
            "src_id": "60",
            "workflow_settings": Object {
              "resource_action_id": "2507",
            },
          },
          "process": true,
          "request_state": "active",
          "request_type": "transformation_plan",
          "requester_id": "1",
          "requester_name": "Administrator",
          "service_order_id": "91",
          "source_id": "80",
          "source_type": "ServiceTemplate",
          "status": "Ok",
          "tenant_id": "1",
          "updated_on": "2018-04-06T12:31:30Z",
          "userid": "admin",
        },
      ],
      "name": "Migration Plan H-0",
      "options": Object {
        "config_info": Object {
          "actions": Array [
            Object {
              "vm_id": "1",
            },
            Object {
              "vm_id": "3",
            },
          ],
          "transformation_mapping_id": "1",
        },
      },
      "scheduleTime": null,
      "status": "Ok",
      "transformation_mapping": Object {
        "transformation_mapping_items": Array [],
      },
    }
  }
/>
`;
